service/matching/matching_engine.go

```diff
@@ -1212,6 +1219,160 @@ func (e *matchingEngineImpl) ReplicateTaskQueueUserData(ctx context.Context, req
+				HandlerError: result.failedWorkerResponse.GetRequest().GetError(),
+			}}, nil
+		}
+
+		return &matchingservice.DispatchNexusTaskResponse{Outcome: &matchingservice.DispatchNexusTaskResponse_Response{
+			Response: result.successfulWorkerResponse.GetRequest().GetResponse(),
+		}}, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (e *matchingEngineImpl) PollNexusTaskQueue(
+	ctx context.Context,
+	req *matchingservice.PollNexusTaskQueueRequest,
+	opMetrics metrics.Handler,
+) (*matchingservice.PollNexusTaskQueueResponse, error) {
+	namespaceID := namespace.ID(req.GetNamespaceId())
+	pollerID := req.GetPollerId()
+	request := req.Request
+	taskQueueName := request.TaskQueue.GetName()
+	e.logger.Debug("Received PollNexusTaskQueue for taskQueue", tag.Name(taskQueueName))
+pollLoop:
+	for {
+		err := common.IsValidContext(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		taskQueue, err := newTaskQueueID(namespaceID, taskQueueName, enumspb.TASK_QUEUE_TYPE_NEXUS)
+		if err != nil {
+			return nil, err
+		}
+
+		// Add frontend generated pollerID to context so taskqueueMgr can support cancellation of
+		// long-poll when frontend calls CancelOutstandingPoll API
+		pollerCtx := context.WithValue(ctx, pollerIDKey, pollerID)
+		pollerCtx = context.WithValue(pollerCtx, identityKey, request.GetIdentity())
+		pollMetadata := &pollMetadata{
+			workerVersionCapabilities: request.WorkerVersionCapabilities,
+		}
+		task, err := e.getTask(pollerCtx, taskQueue, normalStickyInfo, pollMetadata)
+		if err != nil {
+			if err == errNoTasks {
+				return &matchingservice.PollNexusTaskQueueResponse{}, nil
+			}
+			return nil, err
+		}
+
+		e.emitForwardedSourceStats(opMetrics, task.isForwarded(), req.GetForwardedSource())
+
+		if task.isStarted() {
+			// tasks received from remote are already started. So, simply forward the response
+			return task.pollNexusTaskQueueResponse(), nil
+		}
+
+		task.finish(err)
+		if err != nil {
+			continue pollLoop
+		}
+
+		taskToken := &tokenspb.QueryTask{
+			NamespaceId: string(namespaceID),
+			TaskQueue:   taskQueueName,
+			TaskId:      task.nexus.taskID,
+		}
+		serializedToken, _ := e.tokenSerializer.SerializeQueryTaskToken(taskToken)
+		return &matchingservice.PollNexusTaskQueueResponse{
+			Response: &workflowservice.PollNexusTaskQueueResponse{
+				TaskToken: serializedToken,
+				Request:   task.nexus.request.GetRequest(),
+			},
+		}, nil
+	}
+}
+
+func (e *matchingEngineImpl) RespondNexusTaskCompleted(ctx context.Context, request *matchingservice.RespondNexusTaskCompletedRequest, opMetrics metrics.Handler) (*matchingservice.RespondNexusTaskCompletedResponse, error) {
+	resultCh, ok := e.lockableNexusResultMap.get(request.GetTaskId())
+	if !ok {
+		opMetrics.Counter(metrics.RespondNexusTaskFailedPerTaskQueueCounter.GetMetricName()).Record(1)
+		return nil, serviceerror.NewNotFound("nexus task not found or already expired")
+	}
+	resultCh <- &nexusResult{
```

Comment:
Unrelated question: Are these blocking channels or are they sized so this doesn't block?

Reply:
Buffer size of 1.

But I think there's a slight risk that multiple concurrent completions for the same task come in and we will end up with blocked coroutines.

We should "pop" the channel off the map here.

Reply:
fixed (will push soon)